### ISSUE TITLE
Unix time the day before a leap second seems incorrect

### ISSUE THREAD

**Issue by tagliagiulio** [astropy/astropy#17092]
### Description
```python
t0 = Time('1978-12-31T23:59:59', format='isot', scale='utc')
t0.to_value('unix')
```
returns 283996798.0000231  instead of 283996799

in general it seems that all unix time stamp the day before a leap second are one second off.

### Expected behavior

unix epoch of '1978-12-31T23:59:59' should be 283996799

### How to Reproduce
```python
from astropy.time import Time
t0 = Time('1978-12-31T23:59:59', format='isot', scale='utc')
t0.to_value('unix')
```

### Versions

Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.31
Python 3.12.6 (main, Sep 10 2024, 00:05:17) [GCC 9.4.0]
astropy 6.1.4
Numpy 2.1.1
pyerfa 2.0.1.4
Scipy not installed
Matplotlib not installed

**Comment by pllim** [comment]
Unix time does not support leap seconds but I am not an expert here so maybe @mhvk or @taldcroft can clarify. 🙏

**Comment by tagliagiulio** [comment]
@pllim yes correct unix time does not support leap second. However, 1978-12-31T23:59:59 is a non leap second epoch and thus should be supported.

**Comment by mhvk** [comment]
@tagliagiulio - astropy follows the IAU recommendation for leap seconds, which is to spread them out equally over the day. See https://docs.astropy.org/en/latest/api/astropy.time.TimeUnix.html#astropy.time.TimeUnix

This has the disadvantage of breaking exact correspondence with unix time, but the advantage that one can always round-trip from unix time to UTC, etc. It is a choice we had to make, not necessarily the best, but one we cannot undo. Basically, the message should be that one should never use unix time in astronomy - it will just lead to problems. One can use `UnixTai` since that is well defined.

**Comment by eerovaher** [comment]
> astropy follows the IAU recommendation for leap seconds, which is to spread them out equally over the day.

If there is an IAU recommendation for this then it's best we follow it. But there should be a reference to the recommendation in the `TimeUnix` docstring.

**Comment by mhvk** [comment]
I'm basing my comment on the Standards of Fundamental Astronomy (which done by the IAU); see http://www.iausofa.org/sofa_ts_c.pdf, in particular section 4.2 (with more general explanation in 3.5.1).

A PR with links to these would indeed be a good idea. @tagliagiulio - interested?

**Comment by pllim** [comment]
PR = pull request

https://docs.astropy.org/en/latest/index_dev.html

**Comment by fmeynadier** [comment]
Hi,

Quoting http://www.iausofa.org/2023_1011_C/sofa/sofa_ts_c.pdf page 20 (footnote) :

"The quasi-JD that SOFA uses internally to represent UTC is obtained by treating days that end in a leap
second as having a different length from usual, namely 86401s (or 86399s in the case of a negative leap second),
when forming the fractional part. The scheme has no official status outside SOFA, and it is strongly recommended
that applications do not exploit it."

... which looks like an incentive /not/ to do the same, and is certainly not a "IAU recommendation" (which designates a very specific and high-level piece of text, presented at the General Assembly and subject to a vote by all members, which gives it a very high level of authority)... 

However it seems logical that ERFA sticks to what SOFA does as it aims at strict equivalence with it, and then in turn it seems natural for astropy to simply follow up. But we might wonder where the "application" starts... And yes it looks a bit contradictory !

Full disclosure : Giulio and I are working at the BIPM Time department, and I am also a newly elected member of the IAU, participating to A3 commission, though not part of the SOFA board. Leap seconds are (as we all know) a pain to implement and there is no 100% satisfactory implementation, which is one of the reasons why there is a huge collective effort to sort out the matter at the top level: it is very difficult and slow, for a number of not-always-technical reasons... Current horizon is 2035, but if the risk of a "negative leap second" materializes more clearly it may motivate some acceleration to the calendar !

The drawback of spreading the leap second on the whole day is that each fractional MJD time tag will be "wrong" that day (from ~µs error for the first second of the day to ~ 1s). Whether this is better or worse than confining the "wrong" timetags to the last second of the day (for example) depends on the application. There is no defined standard for that, companies operating large server networks all have their own recipe...

Is this the kind of information you would like to see in the docstrings ?

**Comment by mhvk** [comment]
Sorry to have been misleading here! The SOFA guide is really all I saw but clearly I didn't read it well enough to remember more than that I had seen it in some IAU supported guide.

Note that unix time is in some ways a different beast, since it just repeats a leap second, i.e., there is no way to know what second one really refers to. So, for that, the conclusion remains clear: do not use it in astronomy. Instead, convert as quickly as possible to TAI (or TDB) and don't look back...

**Comment by mhvk** [comment]
More concretely, I think it makes sense to perhaps have a small subsection on leap seconds on the main documentation page, to which classes like `TimeUnix` can refer.

**Comment by fmeynadier** [comment]
@mhvk no apologies needed here ! ;-) As you can see we are all in the same boat, trying to make some sense out of contradictory requirements... I can see very well why it has been chosen to follow SOFA on that (even if the footnote discourages that), as it complicates things quite a bit to do otherwise. And with no guarantee that everyone will be happy anyway !

I'll try to come up with a suggestion for the documentation.

And yes until the leap second issue is sorted out by a decision to loosen the UT1-UTC value, applications needing a continuous timescale should probably convert their timetags to TAI as soon as they can.

I like very much the way astropy separates time scales from their representation, and it will be completely straightforward once UTC is "continuous", but until then fractional day representations such as JD and MJD will always be troublesome for UTC around leap seconds insertions...
